Laboratory work #1, Shhanukova Milana - 19FPL1

MilanaShhanukova opened the pull request.

Commit:
Cозданы первые функции

MilanaShhanukova commented:
reopened pull request

Commit:
Новая функция + дебаг

Добавлена новая функция проверки типов и их особенностей для конкорденса, а также проверены функции для тестов.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
-"""
-Lab 1
-A concordance extraction
```
Please, leave lab description and descriptions of the functions as they are.

MilanaShhanukova replied:
Changed. Thanks.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
 
+def checker(con_object, type_ob):
+    if type_ob is list and type(con_object) is list:
+        len_ = len(con_object) > 0
```
It is usually considered a bad style to use built-ins (e.g. `len`, `str`, `list`) as variable names.

MilanaShhanukova replied:
Okay, changed.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
 
 
+def checker(con_object, type_ob):
+    if type_ob is list and type(con_object) is list:
```
There is a short-hand version of this check:
`isinstance(object, type)` - take a look at differences.

MilanaShhanukova replied:
Used, but it does not exclude the process of checking the type the func gets. Thanks

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+
+        if len_ and bull_:
+            return True
+    elif type_ob is str and type(con_object) is str and len(con_object) >= 0:
```
String cannot have negative length.

MilanaShhanukova replied:
Yes, this is a typo. It was meant to be "!=0"

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+
+# токенезирование текста - пунктуация и заглавные
 def tokenize(text: str) -> list:
-    """
```
Same as the first comment.

MilanaShhanukova replied:
done

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+    tokens = []
 
+    for token in text.lower().split():
+        token = "".join([a for a in token if a.isalpha()])
```
`a` is not a pythonic way to name a variable. It will be (and is) rather hard to read and understand what is `a`.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+
+# получение стоп-слов
+def get_stop_words() -> list:
+    with open("stop_words.txt") as file:
```
`open` takes some other arguments like encoding and mode - it is important to pass them so that everyone opens it the same way even on the different operating systems.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
 def remove_stop_words(tokens: list, stop_words: list) -> list:
-    """
-    Removes stop words
-    :param tokens: a list of tokens
```
Same as the first comment.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+# подсчет частотности
 def calculate_frequencies(tokens: list) -> dict:
-    """
-    Calculates frequencies of given tokens
```
Same as the first comment.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+    if checker(tokens, list):
+        frequencies = {token: tokens.count(token) for token in tokens if type(token) is str}
+        return frequencies
+    else:
```
You do not need the `else`-block here. Please explain why.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
 
+# получение top-n популярных слов
 def get_top_n_words(freq_dict: dict, top_n: int) -> list:
-    """
```
Same as the first comment.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
-    """
-    pass
+    if checker(tokens, list):
+        frequencies = {token: tokens.count(token) for token in tokens if type(token) is str}
```
You have `checker` function for type checking, right?

MilanaShhanukova replied:
changed, thanks

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+    return top
 
-def get_concordance(tokens: list, word: str, left_context_size: int, right_context_size: int) -> list:
-    """
```
Same as the first comment.

MilanaShhanukova replied:
changed, thanks

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
-    pass
 
+# получение конкорданса для word
+def get_concordance(tokens: list, word: str, left_n: int, right_n: int) -> list:
```
Please, do not alter the interface of the given functions nor rename the arguments.

MilanaShhanukova replied:
okay, did not know. Changed

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+                    try:
+                        concordance = tokens[index_word - left_n: index_word] + tokens[index_word: index_word + right_context_size + 1]
 
-def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> list:
```
Same as the first comment.

MilanaShhanukova replied:
changed, thanks

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -1,94 +1,121 @@
+    return conco_all
+
+
+def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> list:
```
Please, do not alter the interface of the given functions nor rename the arguments.

Commit:
Исправлены указания.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,41 +4,80 @@
+            return True
+    elif isinstance(con_object, type_ob) and type_ob is str and len(con_object) != 0:
+            return True
+    elif isinstance(con_object, type_ob) and type_ob is int and type_ob is int and con_object >= 0:
```
I guess, one of `type_ob is int` is extra.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,41 +4,80 @@
-    --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    """
-    pass
+      Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
```
Different indentation from the original description.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,41 +4,80 @@
-    e.g. tokens = ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    stop_words = ['the', 'is']
-    --> ['weather', 'sunny', 'man', 'happy']
+       Removes stop words
```
Different indentation from the original description.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,41 +4,80 @@
+    if checker(tokens, list):
+        tokens_clean = [token for token in tokens if token not in stop_words]
+        return tokens_clean
+    else:
```
No need for `else`-block here.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,41 +4,80 @@
-    :return: a dictionary with frequencies
-    e.g. tokens = ['weather', 'sunny', 'man', 'happy']
-    --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
+        Calculates frequencies of given tokens
```
Different indentation from the original description.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -102,13 +193,14 @@ def read_from_file(path_to_file: str) -> str:
 
 
 def sort_concordance(tokens: list, word: str, left_context_size: int, right_context_size: int, left_sort: bool) -> list:
+
```
Extra space.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -102,13 +193,14 @@ def read_from_file(path_to_file: str) -> str:
-    pass
+    with open(path_to_file, "w", encoding="utf-8") as fs:
+        fs.write(" ".join(content))
+
```
Extra space.

Commit:
updated funcs and tested

All of the funcs are updated except the last one which will be done soon.

Commit:
Updated logic

Have not done the whole logic, have question about the interection. Should I create some logic about it or what.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,39 +4,72 @@
+    tokens = []
+
+    for token in text.lower().split():
+        token = "".join([_ for _ in token if _.isalpha()])
```
`_` is not very descriptive, right?

MilanaShhanukova replied:
Changed for a suitable name

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -49,7 +82,13 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
     --> ['happy']
     """
-    pass
+    top = []
```
Do you really need to define it?

MilanaShhanukova replied:
Yes, cause if not I will get None as a result when isinstance(freq_dict, dict) is False, but according to tests I should get [] in such cases

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -69,7 +108,41 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
+    conco_all = []
+    index_word = 0
+
+    if checker(tokens, list) and checker(word, str) \
```
Usually, there are checks for arguments and body function - separated.
And body function is on the same level as checks.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -69,7 +108,41 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
+        if "".join(tokens) != word and word in tokens and normal_concordance:
+            index_word = tokens.index(word, index_word, len(tokens))
+
+            for i in range(tokens.count(word)):
```
`i` is not very descriptive, right?

MilanaShhanukova replied:
okay, I will change it

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -69,7 +108,41 @@ def get_concordance(tokens: list, word: str, left_context_size: int, right_conte
+                try:
+                    concordance = tokens[index_word - left_context_size: index_word] \
+                                  + tokens[index_word: index_word + right_context_size + 1]
+                except IndexError:
```
May be it is better to use some condition here?

MilanaShhanukova replied:
Why is it worse to use try except?

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -87,7 +160,32 @@ def get_adjacent_words(tokens: list, word: str, left_n: int, right_n: int) -> li
+            and left_n >= 0 and right_n >= 0:
+
+        for i in range(tokens.count(word)):
+            if tokens.index(word) == 0:
```
No need to check it explicitly.

MilanaShhanukova replied:
I get a mistake without it

WhiteJaeger replied:
I mean, you can do just:
`if not tokens.index(word)` which results in `True`, just like your line.

Commit:
Updated everything and added new func

Commit:
Update start.py

Commit:
Update target_score.txt

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,39 +4,71 @@
+      """
+    tokens = []
+    for token in text.lower().split():
+        token = "".join([letter for letter in token if letter.isalpha()])
```
Great!

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -49,7 +81,13 @@ def get_top_n_words(freq_dict: dict, top_n: int) -> list:
     """
-    pass
+    top = []
+    if isinstance(freq_dict, dict):
```
Usually, there are checks for input arguments and a function body - separated.
Try to split them and get rid of nesting.

MilanaShhanukova replied:
I don't understand what you mean. We need that check to work with freq_dict, otherwise we get a mistake

WhiteJaeger replied:
You do need that check, yes.
Here what I mean:

```suggestion
    if not isinstance(freq_dict, dict):
        return []
    main body of the function
```

As you can see - they are on the same level - and function body is not nested.
It is much easier to read.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -102,10 +186,8 @@ def read_from_file(path_to_file: str) -> str:
 
 
 def write_to_file(path_to_file: str, content: list):
-    """
```
Leave the function description as it is.

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -125,4 +207,44 @@ def sort_concordance(tokens: list, word: str, left_context_size: int, right_cont
+    conco_all = []
+    index_word = 0
+    left_context_bool, right_context_bool = False, False
+    if isinstance(left_sort, bool):
```
Usually, there are checks for input arguments and a function body - separated.
Try to split them and get rid of nesting.

WhiteJaeger commented on lab_1/target_score.txt:
```diff
@@ -1 +1 @@
-6
\ No newline at end of file
+9
```
There are only 4, 6, 8, 10 available for the sake of automated tests.

WhiteJaeger reviewed (changes requested):
You are almost done!
Just fix the remaining issues.

Commit:
Update main.py

Commit:
Update target_score.txt

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -125,4 +211,44 @@ def sort_concordance(tokens: list, word: str, left_context_size: int, right_cont
+        if checker(tokens, list) and checker(word, str) \
+                and checker(left_context_size, int) and checker(right_context_size, int):
+
+            condition_to_sort = (left_context_size < 0 and left_sort is True) or \
```
No need for explicit check, 
Just use `and left_sort`

MilanaShhanukova replied:
condition_to_sort is that condition that we've discussed. 
If you mean checker for ints, without it I have a bug later.

Commit:
changed

WhiteJaeger committed:
Merge branch 'master' into lab_1

Commit:
checked it

Commit:
for lints changes

Commit:
one small mistake((

Commit:
Update start.py

Commit:
Update start.py

Commit:
Update start.py

Commit:
Update start.py

WhiteJaeger commented on lab_1/main.py:
```diff
@@ -4,39 +4,75 @@
+
+
+def get_stop_words() -> list:
+    with open("stop_words.txt") as file:
```
Define a var `path_to_stop_words` like:
`current_dir = os.path.dirname(os.path.abspath(__file__))`
`path_to_stop_words = os.path.join(current_dir, 'stop_words.txt')`

And open it.

WhiteJaeger commented on lab_1/start.py:
```diff
@@ -1,18 +1,37 @@
+    #  here goes your logic: calling methods from concordance.py
+    tokens = tokenize(data) # токенезирование
+    stop_words = get_stop_words()
+    write_to_file(os.path.join(current_dir, 'lab_1/stop_words.txt'), stop_words)
```
No need for `lab_1`, the following will be enough:
`os.path.join(current_dir, 'stop_words.txt')`

Commit:
Update start.py

Commit:
Update main.py

Commit:
Update start.py

Commit:
Update start.py

Commit:
Update start.py

WhiteJaeger commented:
Open another PR with lab_2.
